tests/matrixstore/test_matrix_usage_providers.py

```diff
@@ -0,0 +1,215 @@
+            command=CommandName.CREATE_LINK.value,
+            args='{"area1": "area1", "area2": "area2", "series": [[1,2,3]]}',
+            index=0,
+            version=7,
+            study_version=study_version,
+        )
+        command_block2 = CommandBlock(
+            study_id=study_id,
+            command=CommandName.CREATE_LINK.value,
+            args='{"area1": "area2", "area2": "area3","series": [[1,2,3]]}',
+            index=0,
+            version=7,
+            study_version=study_version,
+        )
+
+        db.session.add(command_block1)
+        db.session.add(command_block2)
+        db.session.commit()
+
+        matrices_references = command_matrix_usage_provider.get_matrix_usage()
+
+        assert matrices_references == [MatrixReference(matrix_id=matrices_id, use_description=use_description)] * 4
+
+
+def test_constants_matrix_usage_provider(constants_matrix_usage_provider: ConstantsMatrixUsageProvider):
+    constant1 = {"c1": "constant_name1"}
+    constant2 = {"c2": "constant_name2"}
+    constant3 = {"c3": "constant_name3"}
+    constant4 = {"c4": "constant_name4"}
+    constants = [constant1, constant2, constant3, constant4]
+
+    for constant in constants:
+        constants_matrix_usage_provider.matrix_constants.hashes.update(constant)
+
+    constants_reference = constants_matrix_usage_provider.get_matrix_usage()
+    matrix_ref_ids = [ref.matrix_id for ref in constants_reference]
+    matrix_ref_ids.sort()
+
+    constants_id = [valeur for dictio in constants for valeur in dictio.values()]
+
+    constants_id.sort()
+    assert constants_id == matrix_ref_ids
+
+
+def test_dataset_matrix_usage_provider(matrix_service: MatrixService, admin_user):
+    with db():
+        group_repo = GroupRepository()
+        group = group_repo.save(Group(name="groupA", id="groupA"))
+
+        matrix_service.user_service = Mock(spec=LoginService)
+
+        dataset_a = MatrixDataSetUpdateDTO(
+            name="datasetA",
+            public=True,
+            groups=[group.name],
+            created_at=datetime.now(),
+            updated_at=datetime.now(),
+        )
+
+        matrix_m = matrix_service.create(pd.DataFrame([[0]]))
+        matrices = [MatrixInfoDTO(id=matrix_m, name="A")]
+
+        with current_user_context(admin_user):
+            matrix_service.user_service.get_group.return_value = Group(id="foo", name="A")
+
+            dataset_id = matrix_service.create_dataset(dataset_a, matrices)
+
+            use_description = f"Used by dataset {dataset_id.id}"
+
+            matrix_service.usage_providers.append(matrix_service._create_dataset_usage_provider())
+
+            matrices_reference = [
+                reference
+                for usage_prov in matrix_service.usage_providers
+                for reference in usage_prov.get_matrix_usage()
+            ]
+
+            assert matrices_reference == [MatrixReference(matrix_id=matrix_m, use_description=use_description)] * 4
```

Comment:
It's weird that we have 4 of them:
it's because the provider is appended 4 times:
- once every time the creation method is called (2 times)
- once when the matrix service is created
- once in your test

I will remove those duplicates so that we have only one of them